File: Python/AdapterDesign_pattern.py
from abc import ABC, abstractmethod


class IPaymentProcessor(ABC):
    @abstractmethod
    def pay(self) -> None:
        pass


class Application(IPaymentProcessor):
    def pay(self) -> None:
        print("Legacy code")


class StripePayment:
    def make_payment(self) -> None:
        print("Stripe is making payment")


class StripeAdapter(IPaymentProcessor):
    def __init__(self, stripe_payment: StripePayment):
        self.stripe_payment = stripe_payment

    def pay(self) -> None:
        self.stripe_payment.make_payment()


class PayTM:
    def make_payment_paytm(self) -> None:
        print("paytm karo")


class PayTMAdapter(IPaymentProcessor):
    def __init__(self, paytm: PayTM):
        self.paytm = paytm

    def pay(self) -> None:
        self.paytm.make_payment_paytm()


def client_code(process: IPaymentProcessor) -> None:
    process.pay()


if __name__ == "__main__":
    process = StripeAdapter(StripePayment())

    process.pay()

File: Golang/AdapterDesign_pattern.go
package main

import "fmt"

type IPaymentProcessor interface {
	Pay()
}

type Application struct{}

func (a *Application) Pay() {
	fmt.Println("Legacy code")
}

type StripePayment struct{}

func (s *StripePayment) MakePayment() {
	fmt.Println("Stripe is making payment")
}

type StripeAdapter struct {
	stripePayment *StripePayment
}

func NewStripeAdapter(stripePayment *StripePayment) *StripeAdapter {
	return &StripeAdapter{stripePayment: stripePayment}
}

func (s *StripeAdapter) Pay() {
	s.stripePayment.MakePayment()
}

type PayTM struct{}

func (p *PayTM) MakePaymentPaytm() {
	fmt.Println("paytm karo")
}

type PayTMAdapter struct {
	paytm *PayTM
}

func NewPayTMAdapter(paytm *PayTM) *PayTMAdapter {
	return &PayTMAdapter{paytm: paytm}
}

func (p *PayTMAdapter) Pay() {
	p.paytm.MakePaymentPaytm()
}

func clientCode(process IPaymentProcessor) {
	process.Pay()
}

func main() {
	var process IPaymentProcessor = NewStripeAdapter(&StripePayment{})

	process.Pay()
}
